File: app/lib/models/dehalo_parameters.dart
import 'package:json_annotation/json_annotation.dart';

part 'dehalo_parameters.g.dart';

/// Dehalo method options.
@JsonEnum(valueField: 'value')
enum DehaloMethod {
  @JsonValue('DeHalo_alpha')
  dehaloAlpha('DeHalo_alpha', 'DeHalo Alpha'),
  @JsonValue('FineDehalo')
  fineDehalo('FineDehalo', 'Fine Dehalo'),
  @JsonValue('FineDehalo2')
  fineDehalo2('FineDehalo2', 'Fine Dehalo 2'),
  @JsonValue('YAHR')
  yahr('YAHR', 'YAHR'),
  @JsonValue('EdgeCleaner')
  edgeCleaner('EdgeCleaner', 'Edge Cleaner'),
  @JsonValue('Vinverse')
  vinverse('Vinverse', 'Vinverse (de-ghost)'),
  @JsonValue('Vinverse2')
  vinverse2('Vinverse2', 'Vinverse 2 (de-ghost)');

  const DehaloMethod(this.value, this.displayName);
  final String value;
  final String displayName;

  String get description {
    switch (this) {
      case DehaloMethod.dehaloAlpha:
        return 'General purpose dehalo, good for most sources';
      case DehaloMethod.fineDehalo:
        return 'More precise, better edge preservation';
      case DehaloMethod.fineDehalo2:
        return 'Removes the ringing left on sharp edges — run after Fine Dehalo';
      case DehaloMethod.yahr:
        return 'Yet Another Halo Remover - fast and effective';
      case DehaloMethod.edgeCleaner:
        return 'Cleans edge noise and weak halos (aWarpSharp2)';
      case DehaloMethod.vinverse:
        return 'Removes comb/ghost residue left by deinterlacing';
      case DehaloMethod.vinverse2:
        return 'Like Vinverse but keeps more vertical detail';
    }
  }
}

/// Parameters for the dehalo pass.
/// Removes halo artifacts around edges, plus ringing and comb/ghost residue.
///
/// Nullable fields are omitted from the generated script when unset, so
/// havsfunc's own default applies. The non-nullable ones predate that and are
/// always passed — making them nullable would change existing output.
@JsonSerializable()
class DehaloParameters {
  /// Whether this pass is enabled.
  final bool enabled;

  /// Dehalo method to use.
  final DehaloMethod method;

  // --- DeHalo_alpha / FineDehalo parameters ---

  /// Horizontal radius for halo detection (1.0-3.0).
  final double rx;

  /// Vertical radius for halo detection (1.0-3.0).
  final double ry;

  /// Dark halo removal strength (0.0-2.0; above 1.0 overshoots).
  final double darkStr;

  /// Bright halo removal strength (0.0-2.0; above 1.0 overshoots).
  final double brightStr;

  // --- DeHalo_alpha specific ---

  /// Sensitivity to strong halos (havsfunc `lowsens`, 0-100).
  final int? lowSens;

  /// Sensitivity to weak halos (havsfunc `highsens`, 0-100).
  final int? highSens;

  /// Supersampling factor used while removing halos (havsfunc `ss`).
  final double? superSample;

  // --- FineDehalo specific ---

  /// Low threshold for halo mask.
  final int lowThreshold;

  /// High threshold for halo mask.
  final int highThreshold;

  /// Lower limit on how much of a halo may be removed (havsfunc `thlimi`).
  final int? limitLow;

  /// Upper limit on how much of a halo may be removed (havsfunc `thlima`).
  final int? limitHigh;

  /// Contra-sharpening strength applied after dehaloing (havsfunc `contra`).
  final double? contra;

  /// Exclude edges that are too close to each other (havsfunc `excl`).
  final bool? excludeCloseEdges;

  /// How much edge detail to add back (havsfunc `edgeproc`).
  final double? edgeProc;

  // --- YAHR specific ---

  /// Blur amount for YAHR (1-3).
  final int yahrBlur;

  /// Processing depth for YAHR.
  final int yahrDepth;

  // --- EdgeCleaner specific ---

  /// Edge denoising strength (havsfunc `strength`).
  final int? edgeStrength;

  /// Repair the aWarpSharped clip (havsfunc `rep`).
  final bool? edgeRepair;

  /// Repair mode (havsfunc `rmode`: 1 mild, 16/18 keep structure, 17 least halo).
  final int? edgeRepairMode;

  /// Small-particle (star) detection mode (havsfunc `smode`).
  final int? edgeSmallMode;

  /// Remove hot pixels (havsfunc `hot`).
  final bool? edgeHotPixels;

  // --- Vinverse / Vinverse2 specific ---

  /// Sharpening strength applied to the vertically blurred clip (`sstr`).
  final double? vinverseStrength;

  /// Maximum change per pixel (havsfunc `amnt`, 255 = unlimited).
  final int? vinverseAmount;

  /// Process chroma as well as luma (havsfunc `chroma`).
  final bool? vinverseChroma;

  const DehaloParameters({
    this.enabled = false,
    this.method = DehaloMethod.dehaloAlpha,
    this.rx = 2.0,
    this.ry = 2.0,
    this.darkStr = 1.0,
    this.brightStr = 1.0,
    this.lowSens,
    this.highSens,
    this.superSample,
    this.lowThreshold = 50,
    this.highThreshold = 100,
    this.limitLow,
    this.limitHigh,
    this.contra,
    this.excludeCloseEdges,
    this.edgeProc,
    this.yahrBlur = 2,
    this.yahrDepth = 32,
    this.edgeStrength,
    this.edgeRepair,
    this.edgeRepairMode,
    this.edgeSmallMode,
    this.edgeHotPixels,
    this.vinverseStrength,
    this.vinverseAmount,
    this.vinverseChroma,
  });

  DehaloParameters copyWith({
    bool? enabled,
    DehaloMethod? method,
    double? rx,
    double? ry,
    double? darkStr,
    double? brightStr,
    int? lowSens,
    int? highSens,
    double? superSample,
    int? lowThreshold,
    int? highThreshold,
    int? limitLow,
    int? limitHigh,
    double? contra,
    bool? excludeCloseEdges,
    double? edgeProc,
    int? yahrBlur,
    int? yahrDepth,
    int? edgeStrength,
    bool? edgeRepair,
    int? edgeRepairMode,
    int? edgeSmallMode,
    bool? edgeHotPixels,
    double? vinverseStrength,
    int? vinverseAmount,
    bool? vinverseChroma,
  }) {
    return DehaloParameters(
      enabled: enabled ?? this.enabled,
      method: method ?? this.method,
      rx: rx ?? this.rx,
      ry: ry ?? this.ry,
      darkStr: darkStr ?? this.darkStr,
      brightStr: brightStr ?? this.brightStr,
      lowSens: lowSens ?? this.lowSens,
      highSens: highSens ?? this.highSens,
      superSample: superSample ?? this.superSample,
      lowThreshold: lowThreshold ?? this.lowThreshold,
      highThreshold: highThreshold ?? this.highThreshold,
      limitLow: limitLow ?? this.limitLow,
      limitHigh: limitHigh ?? this.limitHigh,
      contra: contra ?? this.contra,
      excludeCloseEdges: excludeCloseEdges ?? this.excludeCloseEdges,
      edgeProc: edgeProc ?? this.edgeProc,
      yahrBlur: yahrBlur ?? this.yahrBlur,
      yahrDepth: yahrDepth ?? this.yahrDepth,
      edgeStrength: edgeStrength ?? this.edgeStrength,
      edgeRepair: edgeRepair ?? this.edgeRepair,
      edgeRepairMode: edgeRepairMode ?? this.edgeRepairMode,
      edgeSmallMode: edgeSmallMode ?? this.edgeSmallMode,
      edgeHotPixels: edgeHotPixels ?? this.edgeHotPixels,
      vinverseStrength: vinverseStrength ?? this.vinverseStrength,
      vinverseAmount: vinverseAmount ?? this.vinverseAmount,
      vinverseChroma: vinverseChroma ?? this.vinverseChroma,
    );
  }

  /// Get a summary string for display.
  String get summary {
    if (!enabled) return 'Off';
    return method.displayName;
  }

  factory DehaloParameters.fromJson(Map<String, dynamic> json) =>
      _$DehaloParametersFromJson(json);
  Map<String, dynamic> toJson() => _$DehaloParametersToJson(this);
}

File: worker/src/models/dehalo_parameters.rs
use serde::{Deserialize, Serialize};

/// Dehalo method options.
#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
pub enum DehaloMethod {
    #[default]
    #[serde(rename = "DeHalo_alpha")]
    DehaloAlpha,
    #[serde(rename = "FineDehalo")]
    FineDehalo,
    /// Second-stage dehalo for the ringing FineDehalo leaves on sharp edges.
    #[serde(rename = "FineDehalo2")]
    FineDehalo2,
    #[serde(rename = "YAHR")]
    Yahr,
    /// Edge denoising / weak dehaloing via aWarpSharp2.
    #[serde(rename = "EdgeCleaner")]
    EdgeCleaner,
    /// Vertical comb/ghost residue removal (the classic post-deinterlace fix).
    #[serde(rename = "Vinverse")]
    Vinverse,
    /// Vinverse variant that preserves more vertical detail.
    #[serde(rename = "Vinverse2")]
    Vinverse2,
}

#[allow(dead_code)]
impl DehaloMethod {
    pub fn as_str(&self) -> &'static str {
        match self {
            DehaloMethod::DehaloAlpha => "DeHalo_alpha",
            DehaloMethod::FineDehalo => "FineDehalo",
            DehaloMethod::FineDehalo2 => "FineDehalo2",
            DehaloMethod::Yahr => "YAHR",
            DehaloMethod::EdgeCleaner => "EdgeCleaner",
            DehaloMethod::Vinverse => "Vinverse",
            DehaloMethod::Vinverse2 => "Vinverse2",
        }
    }

    /// Whether this method takes the shared DeHalo_alpha/FineDehalo radius and
    /// strength arguments.
    pub fn uses_halo_radius(&self) -> bool {
        matches!(self, DehaloMethod::DehaloAlpha | DehaloMethod::FineDehalo)
    }

    /// Whether this method is one of the two Vinverse variants.
    pub fn is_vinverse(&self) -> bool {
        matches!(self, DehaloMethod::Vinverse | DehaloMethod::Vinverse2)
    }
}

/// Parameters for the dehalo pass.
/// Removes halo artifacts around edges, plus ringing and comb/ghost residue.
///
/// Fields typed `Option<T>` are omitted from the generated script when unset,
/// so havsfunc's own default applies. The non-optional ones predate that and
/// are always passed — changing them to Option would change existing output.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DehaloParameters {
    /// Whether this pass is enabled.
    #[serde(default)]
    pub enabled: bool,

    /// Dehalo method to use.
    #[serde(default)]
    pub method: DehaloMethod,

    // --- DeHalo_alpha / FineDehalo parameters ---

    /// Horizontal radius for halo detection (1.0-3.0).
    #[serde(default = "default_rx")]
    pub rx: f64,

    /// Vertical radius for halo detection (1.0-3.0).
    #[serde(default = "default_ry")]
    pub ry: f64,

    /// Dark halo removal strength (0.0-2.0; above 1.0 overshoots).
    #[serde(default = "default_dark_str")]
    pub dark_str: f64,

    /// Bright halo removal strength (0.0-2.0; above 1.0 overshoots).
    #[serde(default = "default_bright_str")]
    pub bright_str: f64,

    // --- DeHalo_alpha specific ---

    /// Sensitivity to strong halos (havsfunc `lowsens`, 0-100).
    #[serde(default)]
    pub low_sens: Option<i32>,

    /// Sensitivity to weak halos (havsfunc `highsens`, 0-100).
    #[serde(default)]
    pub high_sens: Option<i32>,

    /// Supersampling factor used while removing halos (havsfunc `ss`).
    #[serde(default)]
    pub super_sample: Option<f64>,

    // --- FineDehalo specific ---

    /// Low threshold for halo mask.
    #[serde(default = "default_low_threshold")]
    pub low_threshold: i32,

    /// High threshold for halo mask.
    #[serde(default = "default_high_threshold")]
    pub high_threshold: i32,

    /// Lower limit on how much of a halo may be removed (havsfunc `thlimi`).
    #[serde(default)]
    pub limit_low: Option<i32>,

    /// Upper limit on how much of a halo may be removed (havsfunc `thlima`).
    #[serde(default)]
    pub limit_high: Option<i32>,

    /// Contra-sharpening strength applied after dehaloing (havsfunc `contra`).
    #[serde(default)]
    pub contra: Option<f64>,

    /// Exclude edges that are too close to each other (havsfunc `excl`).
    #[serde(default)]
    pub exclude_close_edges: Option<bool>,

    /// How much edge detail to add back (havsfunc `edgeproc`).
    #[serde(default)]
    pub edge_proc: Option<f64>,

    // --- YAHR specific ---

    /// Blur amount for YAHR (1-3).
    #[serde(default = "default_yahr_blur")]
    pub yahr_blur: i32,

    /// Processing depth for YAHR.
    #[serde(default = "default_yahr_depth")]
    pub yahr_depth: i32,

    // --- EdgeCleaner specific ---

    /// Edge denoising strength (havsfunc `strength`).
    #[serde(default)]
    pub edge_strength: Option<i32>,

    /// Repair the aWarpSharped clip (havsfunc `rep`).
    #[serde(default)]
    pub edge_repair: Option<bool>,

    /// Repair mode (havsfunc `rmode`: 1 mild, 16/18 keep structure, 17 least halo).
    #[serde(default)]
    pub edge_repair_mode: Option<i32>,

    /// Small-particle (star) detection mode (havsfunc `smode`).
    #[serde(default)]
    pub edge_small_mode: Option<i32>,

    /// Remove hot pixels (havsfunc `hot`).
    #[serde(default)]
    pub edge_hot_pixels: Option<bool>,

    // --- Vinverse / Vinverse2 specific ---

    /// Sharpening strength applied to the vertically blurred clip (`sstr`).
    #[serde(default)]
    pub vinverse_strength: Option<f64>,

    /// Maximum change per pixel (havsfunc `amnt`, 255 = unlimited).
    #[serde(default)]
    pub vinverse_amount: Option<i32>,

    /// Process chroma as well as luma (havsfunc `chroma`).
    #[serde(default)]
    pub vinverse_chroma: Option<bool>,
}

fn default_rx() -> f64 { 2.0 }
fn default_ry() -> f64 { 2.0 }
fn default_dark_str() -> f64 { 1.0 }
fn default_bright_str() -> f64 { 1.0 }
fn default_low_threshold() -> i32 { 50 }
fn default_high_threshold() -> i32 { 100 }
fn default_yahr_blur() -> i32 { 2 }
fn default_yahr_depth() -> i32 { 32 }

impl Default for DehaloParameters {
    fn default() -> Self {
        Self {
            enabled: false,
            method: DehaloMethod::default(),
            rx: default_rx(),
            ry: default_ry(),
            dark_str: default_dark_str(),
            bright_str: default_bright_str(),
            low_sens: None,
            high_sens: None,
            super_sample: None,
            low_threshold: default_low_threshold(),
            high_threshold: default_high_threshold(),
            limit_low: None,
            limit_high: None,
            contra: None,
            exclude_close_edges: None,
            edge_proc: None,
            yahr_blur: default_yahr_blur(),
            yahr_depth: default_yahr_depth(),
            edge_strength: None,
            edge_repair: None,
            edge_repair_mode: None,
            edge_small_mode: None,
            edge_hot_pixels: None,
            vinverse_strength: None,
            vinverse_amount: None,
            vinverse_chroma: None,
        }
    }
}
